src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java

```diff
@@ -46,79 +62,69 @@ public void init(ServletConfig servletConfig) throws ServletException {
+            cartService.add(cart, productId, parseQuantityRequestParameter(request));
+            redirectToProductDetailsPageWithSuccessMessage(request, response, productId);
             return;
         } catch (ParseException parseException) {
             setAddToCartErrorRequestAttribute(request, "Not a number!");
         } catch (OutOfStockException outOfStockException) {
             setAddToCartErrorRequestAttribute(request,
                     String.format("Not enough items in stock! Available: %d",
-                            outOfStockException.getAvailableQuantity()
-                    )
+                            outOfStockException.getAvailableQuantity())
             );
         }
-        setProductRequestAttribute(request, productDao.getProduct(productId));
-        request.setAttribute("cart", cartService.getCart(httpSession));
-        request.getRequestDispatcher(PRODUCT_DETAILS_JSP).forward(request, response);
-    }
-
-    private void setAddToCartErrorRequestAttribute(HttpServletRequest request, String errorMessage) {
-        request.setAttribute("addToCartError", errorMessage);
+        doGet(request, response);
     }
 
-    private void setProductRequestAttribute(HttpServletRequest request, Product product) {
-        request.setAttribute("product", product);
+    private void redirectToProductDetailsPageWithSuccessMessage(HttpServletRequest request, HttpServletResponse response,
+                                                                Long productId) throws IOException {
+        String redirectPath = String.format("%s/products/%d?message=%s", request.getContextPath(), productId,
+                "Product was added to cart");
+        response.sendRedirect(redirectPath);
     }
 
-    private void setRecentProductsAttribute(HttpServletRequest request) {
-        request.setAttribute("recentProducts", recentProductsService.getRecentProducts(request.getSession()));
+    private int parseQuantityRequestParameter(HttpServletRequest request) throws ParseException {
+        NumberFormat numberFormat = NumberFormat.getInstance(request.getLocale());
+        String quantityString = request.getParameter(QUANTITY_REQUEST_ATTRIBUTE);
+        return numberFormat.parse(quantityString).intValue();
     }
 
-    private void updateRecentProductsSessionAttribute(HttpServletRequest request, Long productId){
-        HttpSession httpSession = request.getSession();
-        recentProductsService.addToRecentProducts(httpSession, productId);
+    private Long getProductIdFromRequestPath(HttpServletRequest request, HttpServletResponse response)
+            throws IOException {
+        Long productId = null;
+        String requestPath = getRequestPath(request);
+        if (requestPath.isEmpty()) {
+            response.sendError(HttpServletResponse.SC_NOT_FOUND);
+        }else{
+            productId = Optional.of(requestPath).map(Long::valueOf).orElse(null);
```

Comment:
requestPath can't be null here as in this case requestPath.isEmpty() will throw an exception before jumping to this lin
so, i don't see any reason to use Optional here,  Long.valueOf(productId) is enough